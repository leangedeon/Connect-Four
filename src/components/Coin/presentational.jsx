import React from "react";
import PropTypes from "prop-types";
import { StyledButton } from './styled.jsx';
const PATH_API = process.env.PATH_API;

export const Coin = ({position, gameId, setLoading, setGame, player}) => {

    const persistMovement = async (gameId, position, player) => {
        const options = {  headers: { 'Content-Type': 'application/json' }, method: 'POST', body: JSON.stringify({'player': player})};
        const data = await fetch(`${PATH_API}/game/${gameId}/move/${position}`, options);
        const dataJSON = await data.json();
        return dataJSON;
    }

    const getGame = async (gameId) => {
        const data = await fetch(`${PATH_API}/game/${gameId}`)
        const dataJSON = await data.json()

        return dataJSON;
    }

    const checkTurn = async (gameId, player) => {
        const data = await getGame(gameId);

        if (data.next !== player && data.state !== 'DONE') {
            return checkTurn(gameId, player);
        }
    }

    const makeMovement = async (gameId, position, player) => {
        setLoading(true);

        const updatedGame = await persistMovement(gameId, position, player);

        setGame(updatedGame);

        await checkTurn(gameId, player);
        setLoading(false);

        const updatedGameAfterPlay = await getGame(gameId);

        setGame(updatedGameAfterPlay);
    }

    return (
        <StyledButton onClick={() => makeMovement(gameId, position, player)}>
            <button type="button" className="btn btn-primary">Put Coin</button>
        </StyledButton>
    )
}

Coin.propTypes = {
    position: PropTypes.number,
    gameId: PropTypes.string,
    setGame: PropTypes.func,
    setLoading: PropTypes.func,
    player: PropTypes.number
}


