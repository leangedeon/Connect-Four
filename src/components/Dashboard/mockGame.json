{
    "dashboard": [
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        },
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        },
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        },
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        },
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        },
        {
            "columns": [
                null,
                null,
                null,
                null,
                null,
                null
            ]
        }
    ],
    "state": "INITIAL"
}