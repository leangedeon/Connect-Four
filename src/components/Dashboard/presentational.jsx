import React, { useEffect } from 'react';
import PropTypes from "prop-types";
import { Row } from "../../components/Row/index.jsx";
import { ListCoins } from "../../components/ListCoins/index.jsx";
import { reverseObject } from '../../utils';
import { Notification } from '../Notification/index.jsx';
import { Status } from '../Status/index.jsx';
const PATH_API = 'http://localhost:3001';

export const Dashboard = ({game, gameId, player, setPlayer, loading, setGame}) => {
    const gameReversed = reverseObject(game.dashboard);

    const createGame = async () => {
        const options = {  headers: { 'Content-Type': 'application/json' }, method: 'POST'};
        const data = await fetch(`${PATH_API}/game`, options);
        const dataJSON = await data.json();
        return dataJSON;
    }

    const getGame = async () => {
        const data = await fetch(`${PATH_API}/game`)
        const dataJSON = await data.json()
        return dataJSON;
    }

    useEffect( async () => {

        if (!gameId) {
            let game = {};

            if(player === 1)  {
                game = await createGame();
            } else {
                game = await getGame();
            }

            setGame(game);
            setPlayer(player);
        }
    }, []);

    return (
        <div>
            <nav className="navbar navbar-expand-lg navbar-dark bg-primary">
                <div className="container-fuild">
                    <a href="/" className="navbar-brand">
                        Connect Four
                    </a>
                </div>
            </nav>
            <main className="bg">
                <div className="container">
                    <div className="row">
                        <h3>Welcome player #{player}</h3>
                        <p className="text-secondary">Game Id {gameId}</p>
                    </div>
                    <Status gameState={game.state}/>
                    
                    <div>

                        {
                            (loading) ? <Notification /> : <ListCoins quantity={6} />
                        }

                        {
                            gameReversed.map((row, index) => (
                                <Row rowData={row} key={`row_${index}`} />
                            ))
                        }
                    </div>
                </div>
            </main>
        </div>
    )

}

Dashboard.propTypes = {
    player: PropTypes.number.isRequired,
    loading: PropTypes.bool.isRequired,
    setPlayer: PropTypes.func.isRequired,
    gameId: PropTypes.string,
    setGame: PropTypes.func.isRequired,    
    game: PropTypes.shape({
        dashboard: PropTypes.object,
        state: PropTypes.string
      })
}
