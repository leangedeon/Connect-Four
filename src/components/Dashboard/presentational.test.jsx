import React from 'react';
import { shallow } from 'enzyme';
import { Dashboard } from './presentational';
import { Notification } from '../Notification/index.jsx';
import { Status } from '../Status/index.jsx';
import { ListCoins } from "../../components/ListCoins/index.jsx";
import mockGame from './mockGame.json';


describe('Dashboard', () => {

    let props = {};
    let wrapper = {};

    describe('Dashboard when we have loading on false', () => {
        
        beforeEach(()=> {    
            props = {
                "game": mockGame,
                "gameId": null,
                "player": 1,
                "setPlayer": jest.fn(),
                "loading": false,
                "setGame": jest.fn()
            };

            wrapper = shallow(<Dashboard {...props} />);
        });

        it('should match snapshot', () => {

            expect(wrapper).toMatchSnapshot();
        });

        it('we must show Status Game always', () => {
            const component = wrapper.find(Status);

            expect(component).toBeTruthy();
        });

        it('if loading false we must show a ListCoins', () => {
            const component = wrapper.find(ListCoins);

            expect(component).toBeTruthy();

            const componentNotRendered = wrapper.find(Notification);

            expect(componentNotRendered.exists()).toBe(false);

        });
    });

    describe('Dashboard when we have loading on false', () => {

        beforeEach(()=> {    
            props = {
                "game": mockGame,
                "gameId": null,
                "player": 1,
                "setPlayer": jest.fn(),
                "loading": true,
                "setGame": jest.fn()
            };

            wrapper = shallow(<Dashboard {...props} />);
        });

        it('if loading true we must show a Notification', () => {

            const component = wrapper.find(Notification);

            expect(component).toBeTruthy();

            const componentNotRendered = wrapper.find(ListCoins);

            expect(componentNotRendered.exists()).toBe(false);

        });
    });
});