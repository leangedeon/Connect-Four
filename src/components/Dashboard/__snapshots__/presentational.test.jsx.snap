// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Dashboard when we have loading on false should match snapshot 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-primary"
  >
    <div
      className="container-fuild"
    >
      <a
        className="navbar-brand"
        href="/"
      >
        Connect Four
      </a>
    </div>
  </nav>
  <main
    className="bg"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <h3>
          Welcome player #
          1
        </h3>
        <p
          className="text-secondary"
        >
          Game Id 
        </p>
      </div>
      <Status
        gameState="INITIAL"
      />
      <div>
        <ListCoins
          quantity={6}
          show="INITIAL"
        />
        <Row
          key="row_0"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
        <Row
          key="row_1"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
        <Row
          key="row_2"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
        <Row
          key="row_3"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
        <Row
          key="row_4"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
        <Row
          key="row_5"
          rowData={
            Object {
              "columns": Array [
                null,
                null,
                null,
                null,
                null,
                null,
              ],
            }
          }
        />
      </div>
    </div>
  </main>
</div>
`;
